aten/src/ATen/core/TensorBase.h

```diff
@@ -636,6 +636,14 @@ class TORCH_API TensorBase {
     return impl_->requires_grad();
   }
 
+  // The Forward AD API functions below are low level and are not to be used by end
+  // users who should use the API provided in torch/csrc/autograd.h
+
+  /// This function returns the forward gradient for this Tensor at the given level.
+  const Tensor& _fw_grad(uint64_t level) const {
```

Comment:
Why was this not here before?
Are there other fw grad methods? If so, this one should be next to them.

Reply:
Not too sure why it wasn't here before. Probably because it wasn't needed anywhere. I've added the other one too now for completeness.

Reply:
We are already using them successfully in https://github.com/pytorch/pytorch/blob/533e72e0a4a7595d157c779d1f001ce78b29f80e/tools/autograd/gen_variable_type.py#L331 and https://github.com/pytorch/pytorch/blob/533e72e0a4a7595d157c779d1f001ce78b29f80e/torch/csrc/autograd/FunctionsManual.cpp#L54 so I don't think we should need them here.

Reply:
In those cases those are Tensors and not TensorBase. Adding this will address the TensorBase case.

Reply:
Ho I forgot about TensorBase, good point!